File: JavaCheatSheet/src/main/java/com/github/jameswdelancey/javaCheatSheet/Part6.java
package com.github.jameswdelancey.javaCheatSheet;

public class Part6 {
    public static void main(String[] args) {
        Hands hands = new Hands();
        hands.report();
        hands = hands.juggle();
        hands.report();
    }
}

class Hands {
    private Item left;
    private Item right;

    public Hands() {
        this.left = new Item("an apple", true);
        this.right = new Item("a banana", true);
    }

    public Hands juggle() {
        System.out.println("Let's juggle!");
        Item air = this.left;
        this.left = this.right;
        this.right = air;
        return this;
    }

    public void report() {
        left.reportItem("Left");
        right.reportItem("Right");
    }
}

class Item {
    private String what;
    private boolean present;

    public Item(String what, boolean present) {
        this.what = what;
        this.present = present;
    }

    public void reportItem(String which) {
        if (present) {
            System.out.println(which + " hand is holding " + what);
        } else {
            System.out.println(which + " hand is not holding anything");
        }
    }
}

File: rustCheatSheet/src/part6.rs
#[allow(unused_imports)]
use Model::Hands;

mod Model {
    pub struct Item {
        what: String,
        present: bool,
    }

    pub struct Hands {
        left: Item,
        right: Item,
    }
    impl Hands {
        pub fn new() -> Hands {
            Hands {
                left: Item {
                    what: "an apple".to_owned(),
                    present: true,
                },
                right: Item {
                    what: "a banana".to_owned(),
                    present: true,
                },
            }
        }

        pub fn juggle(mut self) -> Self {
            println!("Let's juggle!");
            let air = self.left;
            self.left = self.right;
            self.right = air;
            self
        }
        pub fn report(&self) {
            Item::reportItem(&self.left, "Left");
            Item::reportItem(&self.right, "Right");
        }
    }

    impl Item {
        fn reportItem(&self, which: &str) {
            if self.present {
                println!("{} hand is holding {}", which, self.what);
            } else {
                println!("{} hand is not holding anything", which);
            }
        }
    }
}

#[test]
fn main() {
    let mut hands = Hands::new();
    hands.report();
    hands = hands.juggle();
    hands.report();
}
